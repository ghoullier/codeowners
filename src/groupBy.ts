import type { Codeowners } from "./types.ts";

export function groupCodeownersBy(codeowners: Codeowners) {
  return codeowners.reduce(
    ({ byOwner, byPath }, [path, owners]) => {
      const value = byPath.get(path) ?? new Set();
      for (const owner of owners) {
        const paths = byOwner.get(owner) ?? new Set();
        paths.add(path);
        byOwner.set(owner, paths);
        value.add(owner);
      }
      byPath.set(path, value);
      return { byOwner, byPath };
    },
    {
      byOwner: new Map<string, Set<string>>(),
      byPath: new Map<string, Set<string>>(),
    },
  );
}
