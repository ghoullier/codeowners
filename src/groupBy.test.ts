import { describe, expect, it } from "bun:test";

import { groupCodeownersBy } from "./groupBy.js";

describe("groupBy()", () => {
  it("should group by paths", () => {
    expect(
      groupCodeownersBy([
        ["apps/", ["@org/team1"]],
        ["packages/", ["@org/team2"]],
      ]),
    ).toMatchSnapshot();
  });
  it("should group by owners", () => {
    expect(
      groupCodeownersBy([
        ["apps/", ["@org/team1", "@org/team3"]],
        ["packages/", ["@org/team2"]],
        ["shared/", ["@org/team1", "@org/team2"]],
      ]),
    ).toMatchSnapshot();
  });
});
