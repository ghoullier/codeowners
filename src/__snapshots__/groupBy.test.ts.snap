// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`groupBy() should group by paths 1`] = `
{
  "byOwner": 
Map {
    "@org/team1" =>     
Set {
      "apps/",
    }
,
    "@org/team2" =>     
Set {
      "packages/",
    }
,
  }
,
  "byPath": 
Map {
    "apps/" =>     
Set {
      "@org/team1",
    }
,
    "packages/" =>     
Set {
      "@org/team2",
    }
,
  }
,
}
`;

exports[`groupBy() should group by owners 1`] = `
{
  "byOwner": 
Map {
    "@org/team1" =>     
Set {
      "apps/",
      "shared/",
    }
,
    "@org/team3" =>     
Set {
      "apps/",
    }
,
    "@org/team2" =>     
Set {
      "packages/",
      "shared/",
    }
,
  }
,
  "byPath": 
Map {
    "apps/" =>     
Set {
      "@org/team1",
      "@org/team3",
    }
,
    "packages/" =>     
Set {
      "@org/team2",
    }
,
    "shared/" =>     
Set {
      "@org/team1",
      "@org/team2",
    }
,
  }
,
}
`;
